Add total -> subtotal calculation correction & fix hard-coded plant-in-service table name

cmgosnell opened the pull request:
# Overview

Closes #3404 and closes on #3309.

What problem does this address?

### What did you change?
- add total -> sub-dimension corrections
- rename `subtotal` to `subdimension` to keep it consistent
- consolidated how we are reconciling table calcs via `reconcile_one_type_of_table_calculations`
- enable multiple dims in the sub-dimension calculation reconciliation: did this by adding `subdimension_merge_validation` to enable a `many_to_many` merge for more than one dimension column
- add unit tests to explode multi-dimension correction records
- enable reported parent w/ # -> null calculation component correction
   - this required a bunch of increases in the calculation metric checking badness
- explicitly instead of implicitly add the standard `_correction` and the new `_subdimension_correction` records to the calculation components table

### What is still a problem?
#### for the `depreciation_by_function` table subtotal corrections are not in `_core_ferc1_xbrl__calculation_components`: SOLVED
- we are ending up with ~200 correction records almost all for 2021 in the core table which is exactly what we'd expect.
- but table has one factoid. and no inherent calculations with one factoid. the calculations are all total -> subtotals. rn there are total -> subtotal calculations happening for this table in its transform. 
- the way we compile the parts for the calc components table in each table transform is slightly different than how we do it in `_core_ferc1_xbrl__calculation_components` and we need some way to make sure the subtotal ->total calcs and either corrections end up in the final table:
  - option: standardize moar between the transforms and the core calc component table! this feels hard bc there are different needs/inputs into the transforms calc components and the core all table calc component
  - option: cache and reuse each transformed table-level calc component and then compile them together and deal with the inter-table calcs then.
  - option (preferred): be more explicit about adding corrections in the core calc components *at the end*. I think rn we are missing these corrections bc they are semi-haphazardly being added during `make_xbrl_factoid_dimensions_explicit`, but this table has so calc comps until `infer_intra_factoid_totals`
#### Does the current setup for many:many result in double counting in the rate base table??
- i looks like the answer here is no. I've checked several data points (grabbing a utility/year and checking through the process), but i'd like to add some output unit tests for it as well that are the follow up to the transform unit tests.
- Why? well because `infer_intra_factoid_totals` has no "total"s in the children.... that means when there are mutli-dimensions, the calculation for the total/total (parents) record is connected to sub-dims/sub-dims (children). there are total/sub-dim & sub-dim/total calculations, but those are not actually connected to the total/total calculation are are thus pruned from the tree.

## remaining todos:
```[tasklist]
 - [x] convert param notes to TODOS (I decided to convert them all to `# worst {thing}: {worst one}`, bc its not realllly a todo, its mostly a helpful lil "hey if your looking into this maybe start w/ this worst guy"/"why is 2021 consistently bad across the tables"
 - [x] check to see if there are null total/total records with non-null total/sub-dim records (& thus are we loosing data)
- [x] convert the one unit test into many tests
- [x] add unit test to ensure dropping the sub-dim/total calcs (re:double counting!)
 - [ ] make an issue about standardizing `row_type_xbrl` instead of assigning `is_calc`
```

# Testing

How did you make sure this worked? How can a reviewer verify this?

```[tasklist]
# To-do list
- [ ] Ensure docs build, unit & integration tests, and test coverage pass locally with `make pytest-coverage` (otherwise the merge queue may reject your PR)
- [ ] For significant ETL changes, ensure the full ETL runs locally
- [ ] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [ ] If updating analyses or data processing functions: make sure to update or write data validation tests
- [ ] Update the [release notes](../docs/release_notes.rst): reference the PR and related issues.
- [ ] Review the PR yourself and call out any questions or issues you have
```

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -4593,7 +4644,7 @@ def convert_xbrl_metadata_json_to_df(
             {
                 "name": "electric_plant_in_service_and_completed_construction_not_classified_electric",
                 "weight": 1.0,
-                "source_tables": ["plant_in_service_ferc1"],
+                "source_tables": ["core_ferc1__yearly_plant_in_service_sched204"],
                 "utility_type": "electric",
             }
```
this old table name was resulting in the vast majority of the $s from this table not showing up in the rate base table

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -963,7 +966,7 @@ def reconcile_table_calculations(
         and ``rel_diff``.
     """
     # If we don't have this value, we aren't doing any calculation checking:
-    if params.column_to_check is None or calculation_components.empty:
+    if params.column_to_check is None:
```
previously the `depreciation_by_function` table was entirely getting skipped over because it has no calc components prior to inferring the total to sub-dim calcs via `_add_intra_table_calculation_dimensions`

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -1054,32 +1105,6 @@ def _calculation_components_subtotal_calculations(
     return calc_comps_w_totals
 
 
-def _check_subtotal_calculations(
```
now both the "standard" calcs and the subtotal calcs are being checked via `reconcile_one_type_of_table_calculations`

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -1173,18 +1201,19 @@ def calculate_values_from_components(
             pd.merge(
                 calculation_components,
                 data,
-                validate="one_to_many",
+                validate=calc_to_data_merge_validation,
```
we were not able to test the `depreciation_by_function` table because it has two dimension columns and child calc components *necessarily & correctly* are part of more than one parent calculation.

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -1475,7 +1515,7 @@ def apply_metric(self: Self, df: pd.DataFrame) -> pd.Series:
             df[df.row_type_xbrl == "calculated_value"]
             .groupby(self.groupby_cols())
-            .apply(self.metric)
+            .apply(self.metric, include_groups=False)
```
this was to remove a future warning and preserve existing behavior

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -4546,7 +4623,7 @@ def convert_xbrl_metadata_json_to_df(
             }
             for new_fact in ["megawatt_hours_purchased", "purchased_mwh"]
         ]
-        new_facts = pd.DataFrame(facts_to_add).convert_dtypes()
+        new_facts = pd.DataFrame(facts_to_add).astype(tbl_meta.dtypes, errors="ignore")
```
these were to remove future warnings about concat-ing fully null columns

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -6428,26 +6510,8 @@ def make_xbrl_factoid_dimensions_explicit(
             the parental dimensions or the child dimensions.
     """
     logger.info(f"Adding {dimensions=} into calculation component table.")
-    # even though we don't actually get correction records for all of the
-    # factiods, we are still going to force them to exist here so all of the
-    # downstream processes have them all. All table dim facts get correction
-    # records regardless of whether they are calculated values.
-    # The non-calcualted _correction records will be dropped in the left
-    # merge below while assigning df_w_implied_dims
-    non_correction_mask = ~table_dimensions_ferc1.xbrl_factoid.str.endswith(
-        "_correction"
-    )
-    table_dimensions_ferc1 = pd.concat(
-        [
-            table_dimensions_ferc1[non_correction_mask],
-            (
-                table_dimensions_ferc1[non_correction_mask].assign(
-                    xbrl_factoid=lambda x: x.xbrl_factoid + "_correction"
-                )
-            ),
-        ]
-    )
```
this is now all happening in `add_calculation_component_corrections`

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -6677,6 +6739,55 @@ def infer_intra_factoid_totals(
     return calcs_with_totals
 
 
+def add_calculation_component_corrections(
```
me q: rn this is only being used to compile the big mutli-table calc component table. can i use this in the table transforms as well?

cmgosnell commented on src/pudl/transform/params/ferc1.py:
many of the calculation metrics got worse because we are now including calculations when there is a null on either the calculated or reported side of the calculations.

cmgosnell commented on test/unit/transform/ferc1_test.py:
```diff
@@ -548,6 +556,207 @@ def test_calculate_values_from_components():
     )
 
 
+def test_many_to_many_subdimensions_calculate_values_from_components():
```
there are effectively 4 different unit tests in here all using the same lil helper `get_reconcilded_data` to build and poop out the corrected calculations. This could easily be broken up into 4 unit tests. i kinda think it is easier to read all in one place bc it is 4 iterations on the same idea. but very open to rejiggering.

e-belfer commented on src/pudl/transform/params/ferc1.py:
```diff
@@ -3551,11 +3573,16 @@
+                        "error_frequency": 0.23,
+                        "relative_error_magnitude": 1.0,
+                        "null_calculated_value_frequency": 1.0,
+                    },  # 331
```
`TODO: Investigate utility 331`

cmgosnell commented on test/unit/transform/ferc1_test.py:
```diff
@@ -548,6 +556,207 @@ def test_calculate_values_from_components():
+    )
+    out2 = get_reconcilded_data(data2)
+    assert (
+        out2.loc[out2.row_type_xbrl == "subdimension_correction", "value"].to_numpy()[0]
```
add another assertion for only one correction record

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -5713,6 +5713,63 @@ def transform_main(self, df: pd.DataFrame) -> pd.DataFrame:
         return df
 
+    def transform_end(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Run standard :meth:`Ferc1AbstractTableTransformer.transform_end` plus a data validation step.
```
this was much more clear! i made one tweak to this which was in your first e.g.: "(e.g. `plant_status == "total"`)" instead of "in_service" there.

cmgosnell commented:
great catch on the `subdimension_correction` not in the enum! I added `subtotal_correction` into the enum in the schema and then i up and changed it to subdimension bc that was more consistent w/ the code. will change now.

cmgosnell commented:
i just re-ran the fast etl which i hadn't done in a minute and noticed there are a few tables that need small adjustments to their calc metric tolerances. i'll update and push momentarily

e-belfer reviewed (approved):
Everything ran as expected and looks good to me.
